linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/BinaryWriter.scala

```diff
@@ -0,0 +1,660 @@
+  private final val SectionCustom = 0x00
+  private final val SectionType = 0x01
+  private final val SectionImport = 0x02
+  private final val SectionFunction = 0x03
+  private final val SectionTable = 0x04
+  private final val SectionMemory = 0x05
+  private final val SectionGlobal = 0x06
+  private final val SectionExport = 0x07
+  private final val SectionStart = 0x08
+  private final val SectionElement = 0x09
+  private final val SectionCode = 0x0A
+  private final val SectionData = 0x0B
+  private final val SectionDataCount = 0x0C
+  private final val SectionTag = 0x0D
+
+  private final class Buffer {
+    private var buf: Array[Byte] = new Array[Byte](1024 * 1024)
+    private var size: Int = 0
+
+    private def ensureCapacity(capacity: Int): Unit = {
+      if (buf.length < capacity) {
+        val newCapacity = Integer.highestOneBit(capacity) << 1
+        buf = java.util.Arrays.copyOf(buf, newCapacity)
+      }
+    }
+
+    def currentGlobalOffset: Int = size
+
+    def result(): Array[Byte] =
+      java.util.Arrays.copyOf(buf, size)
+
+    def byte(b: Byte): Unit = {
+      val newSize = size + 1
+      ensureCapacity(newSize)
+      buf(size) = b
+      size = newSize
+    }
+
+    def rawByteArray(array: Array[Byte]): Unit = {
+      val newSize = size + array.length
+      ensureCapacity(newSize)
+      System.arraycopy(array, 0, buf, size, array.length)
+      size = newSize
+    }
+
+    def boolean(b: Boolean): Unit =
+      byte(if (b) 1 else 0)
+
+    def u32(value: Int): Unit = unsignedLEB128(Integer.toUnsignedLong(value))
+
+    def s32(value: Int): Unit = signedLEB128(value.toLong)
+
+    def i32(value: Int): Unit = s32(value)
+
+    def s33OfUInt(value: Int): Unit = signedLEB128(Integer.toUnsignedLong(value))
+
+    def u64(value: Long): Unit = unsignedLEB128(value)
+
+    def s64(value: Long): Unit = signedLEB128(value)
+
+    def i64(value: Long): Unit = s64(value)
+
+    def f32(value: Float): Unit = {
+      val bits = java.lang.Float.floatToIntBits(value)
+      byte(bits.toByte)
+      byte((bits >>> 8).toByte)
+      byte((bits >>> 16).toByte)
+      byte((bits >>> 24).toByte)
+    }
+
+    def f64(value: Double): Unit = {
+      val bits = java.lang.Double.doubleToLongBits(value)
+      byte(bits.toByte)
+      byte((bits >>> 8).toByte)
+      byte((bits >>> 16).toByte)
+      byte((bits >>> 24).toByte)
+      byte((bits >>> 32).toByte)
+      byte((bits >>> 40).toByte)
+      byte((bits >>> 48).toByte)
+      byte((bits >>> 56).toByte)
+    }
+
+    def vec[A](elems: Iterable[A])(op: A => Unit): Unit = {
+      u32(elems.size)
+      for (elem <- elems)
+        op(elem)
+    }
+
+    def opt[A](elemOpt: Option[A])(op: A => Unit): Unit =
+      vec(elemOpt.toList)(op)
+
+    def name(s: String): Unit =
+      name(UTF8String(s))
+
+    def name(utf8: UTF8String): Unit = {
+      val len = utf8.length
+      u32(len)
+      var i = 0
+      while (i != len) {
+        byte(utf8(i))
+        i += 1
+      }
```

Comment:
With ByteBuffers, we could improve this by adding a `writeTo` method on `UTF8String`.